src/message/dkim.rs

```diff
@@ -0,0 +1,318 @@
+            },
+        }
+    }
+}
+
+/// Create a Headers struct with a Dkim-Signature Header created from given parameters
+fn dkim_header_format(
+    domain: String,
+    selector: String,
+    canon: DkimCanonicalization,
+    timestamp: String,
+    headers_list: String,
+    body_hash: String,
+    signature: String,
+) -> Headers {
+    let mut headers = Headers::new();
+    let header_name = match canon.header {
+        DkimCanonicalizationType::Simple => HeaderName::new_from_ascii_str("DKIM-Signature"),
+        DkimCanonicalizationType::Relaxed => HeaderName::new_from_ascii_str("dkim-signature"),
+    };
+    headers.append_raw(header_name, format!("v=1; a=rsa-sha256; d={domain}; s={selector}; c={canon}; q=dns/txt; t={timestamp}; h={headers_list}; bh={body_hash}; b={signature}",domain=domain, selector=selector,canon=canon,timestamp=timestamp,headers_list=headers_list,body_hash=body_hash,signature=signature));
+    headers
+}
+
+/// Canonicalize the body of an email
+fn dkim_canonicalize_body(body: &[u8], canonicalization: DkimCanonicalizationType) -> String {
+    let body = std::str::from_utf8(body).unwrap();
+    let re = Regex::new("(\r\n)+$").unwrap();
+    match canonicalization {
+        DkimCanonicalizationType::Simple => re.replace(body, "\r\n").to_string(),
+        DkimCanonicalizationType::Relaxed => {
+            let re_double_space = Regex::new("[\\t ]+").unwrap();
+            let body = re_double_space.replace_all(body, " ").to_string();
+            let re_space_eol = Regex::new("[\t ]\r\n").unwrap();
+            let body = re_space_eol.replace_all(&body, "\r\n").to_string();
+            re.replace(&body, "\r\n").to_string()
+        }
+    }
+}
+
+/// Canonicalize the value of an header
+fn dkim_canonicalize_header_value(
+    value: &str,
+    canonicalization: DkimCanonicalizationType,
+) -> String {
+    match canonicalization {
+        DkimCanonicalizationType::Simple => value.to_string(),
+        DkimCanonicalizationType::Relaxed => {
+            let re = Regex::new("\r\n").unwrap();
+            let value = re.replace_all(value, "").to_string();
+            let re = Regex::new("[\\t ]+").unwrap();
+            format!("{}\r\n", re.replace_all(&value, " ").to_string().trim_end())
+        }
+    }
+}
+
+/// Canonicalize signed headers passed as headers_list among mail_headers using canonicalization
+fn dkim_canonicalize_headers(
+    headers_list: Vec<String>,
+    mail_headers: &Headers,
+    canonicalization: DkimCanonicalizationType,
+) -> String {
+    let mut signed_headers = Headers::new();
+    let mut signed_headers_relaxed = String::new();
+    for h in headers_list.into_iter() {
+        let h = match canonicalization {
+            DkimCanonicalizationType::Simple => h,
+            DkimCanonicalizationType::Relaxed => {
+                let mut ret = String::new();
+                ret.push_str(&h);
+                ret.to_lowercase()
+            }
```

Comment:
This has multiple unnecessary copies.

- `h` is owned so there is no need to copy it. 
- `to_lowercase` copies again.

If we only need ASCII lowercase we can do this with no copies.

```suggestion
            DkimCanonicalizationType::Relaxed => {
                h.make_ascii_lowercase();
                h
            }
```

Otherwise 1 copy. (Rust std doesn't provide an easy way to do this without a copy)

```suggestion
            DkimCanonicalizationType::Relaxed => h.to_lowercase(),
```